### PR TITLE
Feature/pairwise similarity

### PR DESCRIPTION
**Pull request by YoussefAch**
### Description
This PR adds two new transformer/layer pairs to enable computing the maximum cosine similarity between an embedding and a **list** of candidate embeddings. From my understanding, the existing implementation of cosine similarity allow only for similarity between two embeddings.

**1. PairwiseCosineSimilarityTransformer / PairwiseCosineSimilarityLayer**
- Takes two inputs: a query embedding (Array[Float] of size D) and a flat array of candidate embeddings (Array[Float] of size N*D)
- Returns an Array[Float] of size N containing the cosine similarity between the query and each candidate embedding

**2. ArrayReduceMaxTransformer / ArrayReduceMaxLayer**
- Takes a single Array[Float] column and reduces it to a scalar by taking the maximum element (F.array_max in Spark, tf.reduce_max(..., axis=-1) in TF)
- Accepts a configurable defaultValue (default 0.0) returned when the array is null or empty

**3. Unit tests for all introduced modules**

- [x] There is a new entry (alphabetical order) in the README table describing the new layer/transformer

### Keras Layer Checklist

- [x] The new Keras layer extends [BaseLayer](../blob/main/src/kamae/tensorflow/layers/base.py)
- [x] The `_call` method has been implemented in the new layer.
- [x] The `compatible_dtypes` property is defined in the new layer.
- [x] The new layer is decorated with `@tf.keras.utils.register_keras_serializable(package=kamae.__name__)`.
- [x] The new layer takes a `name`, `input_dtype`, and `output_dtype` as arguments to the constructor and that this is passed to the super constructor.
- [x] The Keras layer is serializable. I have implemented the `get_config` method.
- [x] There are unit tests of the new layer. 
- [x] There is a specific test of layer serialisation added [here](../blob/main/tests/kamae/tensorflow/test_layer_serialisation.py).
- [x] The new layer is imported in the [__init__.py](../blob/main/src/kamae/tensorflow/layers/__init__.py) file in the `layers` directory.

### Spark Transformer/Estimator Checklist

- [x] The new Spark Transformer extends [BaseTransformer](../blob/main/src/kamae/spark/transformers/base.py).
- [x] If the new transform needs a fit method, a Spark Estimator has been implemented that extends [BaseEstimator](../blob/main/src/kamae/spark/estimators/base.py).
- [x] The instructions in the above docs page have been followed for the `__init__` and `setParams` methods.
- [x] The transformer uses one of the input/output mixin classes from [base.py](../blob/main/src/kamae/spark/params/base.py).
- [x] If the new transformer requires more parameters that would need to be serialised to the Spark ML pipeline, there is a implemented parameter class by extending the `Params` class [here](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.ml.param.Params.html).
- [x] The `compatible_dtypes` property has been implemented to specify the input/output data types that my transformer/estimator supports.
- [x] A Keras subclassed layer is returned in the transformer's `get_tf_layer` method.
- [x] There are unit tests of the new transform. In particular, there are parity tests between the Spark and Keras implementations.
- [x] The new transformer/estimator is imported in the [__init__.py](../blob/main/src/kamae/spark/transformers/__init__.py) file in the `transformers`/`estimators` directory.

### CODE REVIEW

**Comment by georyetti** [comment]
Hey @YoussefAch - raw changes look good. Can you fix the linting/formatting failing checks? 

Also can we add a parity test from Spark -> Tensorflow? You can see in the existing transformer tests how we do this for other transforms.

**Comment by georyetti** [comment]
Happy to get this in before we do Keras 3 migration as that's a whopper

**Comment by YoussefAch** [comment]
Cheers! I'll push the new changes asap!

**Comment by YoussefAch** [comment]
@georyetti Added parity tests and fixed the formatting issues (hopefully)

**Comment by georyetti** [comment]
Test are good now - but the license header check is failing. If you setup the precommit locally it will fix this for you

**Comment by YoussefAch** [comment]
Normally I set the pre-commit and it was fine, let me check this

**Review by georyetti — APPROVED** [review]
LGTM

**Review by mandrecki — APPROVED** [review]
